File: patterns/behavioral/chain-of-responsibility.js
/**
 * Chain of Responsibility Pattern - JavaScript
 * 
 * Passes requests along a chain of handlers. Each handler decides either to
 * process the request or pass it to the next handler in the chain.
 * 
 * Use cases:
 * - Request processing pipelines
 * - Event handling
 * - Middleware
 * - Validation chains
 */

// Handler interface
class Handler {
    constructor() {
        this.next = null;
    }
    
    setNext(handler) {
        this.next = handler;
        return handler;
    }
    
    handle(request) {
        if (this.next) {
            return this.next.handle(request);
        }
        return null;
    }
}

// Concrete handlers
class AuthenticationHandler extends Handler {
    handle(request) {
        if (request.user && request.user.authenticated) {
            console.log('Authentication: OK');
            return super.handle(request);
        }
        console.log('Authentication: FAILED');
        return 'Authentication failed';
    }
}

class AuthorizationHandler extends Handler {
    handle(request) {
        if (request.user && request.user.role === 'admin') {
            console.log('Authorization: OK');
            return super.handle(request);
        }
        console.log('Authorization: FAILED');
        return 'Authorization failed';
    }
}

class ValidationHandler extends Handler {
    handle(request) {
        if (request.data && request.data.length > 0) {
            console.log('Validation: OK');
            return super.handle(request);
        }
        console.log('Validation: FAILED');
        return 'Validation failed';
    }
}

class RequestProcessor extends Handler {
    handle(request) {
        console.log('Processing request:', request);
        return 'Request processed successfully';
    }
}

// Example usage
console.log('=== Chain of Responsibility Pattern ===\n');

// Build chain
const auth = new AuthenticationHandler();
const authz = new AuthorizationHandler();
const validation = new ValidationHandler();
const processor = new RequestProcessor();

auth.setNext(authz).setNext(validation).setNext(processor);

// Test requests
console.log('Request 1 (authenticated, admin, valid data):');
const request1 = {
    user: { authenticated: true, role: 'admin' },
    data: 'some data'
};
console.log('Result:', auth.handle(request1));

console.log('\nRequest 2 (not authenticated):');
const request2 = {
    user: { authenticated: false },
    data: 'some data'
};
console.log('Result:', auth.handle(request2));

console.log('\nRequest 3 (authenticated, but not admin):');
const request3 = {
    user: { authenticated: true, role: 'user' },
    data: 'some data'
};
console.log('Result:', auth.handle(request3));

module.exports = { Handler, AuthenticationHandler, AuthorizationHandler, ValidationHandler, RequestProcessor };

File: patterns/behavioral/chain-of-responsibility.py
"""
Chain of Responsibility Pattern - Python

Passes requests along a chain of handlers. Each handler decides either to
process the request or pass it to the next handler in the chain.

Use cases:
- Request processing pipelines
- Event handling
- Middleware
- Validation chains
"""

from abc import ABC, abstractmethod


# Handler interface
class Handler(ABC):
    def __init__(self):
        self.next = None
    
    def set_next(self, handler):
        self.next = handler
        return handler
    
    @abstractmethod
    def handle(self, request):
        if self.next:
            return self.next.handle(request)
        return None


# Concrete handlers
class AuthenticationHandler(Handler):
    def handle(self, request):
        if request.get('user') and request['user'].get('authenticated'):
            print("Authentication: OK")
            return super().handle(request)
        print("Authentication: FAILED")
        return "Authentication failed"


class AuthorizationHandler(Handler):
    def handle(self, request):
        if request.get('user') and request['user'].get('role') == 'admin':
            print("Authorization: OK")
            return super().handle(request)
        print("Authorization: FAILED")
        return "Authorization failed"


class ValidationHandler(Handler):
    def handle(self, request):
        if request.get('data') and len(request['data']) > 0:
            print("Validation: OK")
            return super().handle(request)
        print("Validation: FAILED")
        return "Validation failed"


class RequestProcessor(Handler):
    def handle(self, request):
        print(f"Processing request: {request}")
        return "Request processed successfully"


# Example usage
if __name__ == "__main__":
    print("=== Chain of Responsibility Pattern ===\n")
    
    # Build chain
    auth = AuthenticationHandler()
    authz = AuthorizationHandler()
    validation = ValidationHandler()
    processor = RequestProcessor()
    
    auth.set_next(authz).set_next(validation).set_next(processor)
    
    # Test requests
    print("Request 1 (authenticated, admin, valid data):")
    request1 = {
        'user': {'authenticated': True, 'role': 'admin'},
        'data': 'some data'
    }
    print("Result:", auth.handle(request1))
    
    print("\nRequest 2 (not authenticated):")
    request2 = {
        'user': {'authenticated': False},
        'data': 'some data'
    }
    print("Result:", auth.handle(request2))
    
    print("\nRequest 3 (authenticated, but not admin):")
    request3 = {
        'user': {'authenticated': True, 'role': 'user'},
        'data': 'some data'
    }
    print("Result:", auth.handle(request3))

